kernel/cClosure.ml

```diff
@@ -1311,7 +1311,7 @@ module FNativeEntries =
 
     let mkArray env u t ty =
       check_array env;
-      { mark = Whnf; term = FArray (u,t,ty)}
+      { mark = Cstr; term = FArray (u,t,ty)}
```

Comment:
The documentation states that "Whnf means we reached the head normal form and that it cannot create a redex when substituted". You seem to think that substituting an array can create a redex. This does not make sense to me. No array can ever be substituted on the left of an application.

Reply:
Do we not substitute on the right?

Reply:
On the right of non-primitive operation, it cannot create a redex either. So, the only remaining case is the right of a primitive operation. But primitive operations have their own zipper, for which there is no right side, as the zipper carries all their arguments. So, there is no substitution on the right for primitive types. (But I am willing to be proven wrong, as this code is quite confusing.)

Reply:
I am not 100% confident that this can happen, but it is a safety measure that cannot create incompleteness, at best we just reduce too much by missing an optimization. A hypothetical scenario where this could happen is some case where, after reducing a partially applied primitive, an at-distance update refolds it indirectly and wrongly treats the right application as a Whnf. In any case I'd rather err on the safe side and make arrays behave similarly to primitive projections for uniformity.

Reply:
(Note also that the other construction of a FArray in knht also marks it as Cstr, so at the very least we should be uniform.)